fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/LineageFactory.java

```diff
@@ -0,0 +1,74 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+package com.starrocks.sql.optimizer.rule.transformation.materialization;
+
+import com.google.common.collect.Maps;
+import com.starrocks.sql.optimizer.OptExpression;
+import com.starrocks.sql.optimizer.OptExpressionVisitor;
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.base.ColumnRefSet;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.Map;
+import java.util.stream.Collectors;
+
+// get all projection map under root
+public class LineageFactory {
```

Comment:
why do need the Lineage?

Reply:
I rewrite the output columns and predicates by utilizing mapping between tables and table's columns.
I want to normalize all output expressions and predicates by base table's columns. So here I create Lineage to trace the mapping of output expression and predicates. Here is an example, there is a mv tree:
Projection - Filter - Scan.
Scan has a column col1, and has a projection col5 -> col1 + 1;
Projection has a project col 6 -> abs(col5)

Then there is a query, the output column is abs(col1 + 1), I should judge whether query's abs(col1 + 1) can be rewritten by col6 of mv, I should collect the lineage and rewrite col6 by lineage to abs(col1 + 1), then I can judge query's abs(col1 + 1) can be rewritten by col6.